src/main/scala/is/hail/distributedmatrix/HailBlockMatrix.scala

```diff
@@ -0,0 +1,587 @@
+
+      def :+(v: Array[Double]): M =
+        l.vectorAddToEveryColumn(v)
+      def :*(v: Array[Double]): M =
+        l.vectorPointwiseMultiplyEveryColumn(v)
+
+      def --*(v: Array[Double]): M =
+        l.vectorPointwiseMultiplyEveryRow(v)
+    }
+    implicit class ScalarShim(l: Double) {
+      def +(r: M): M =
+        r.scalarAdd(l)
+      def -(r: M): M = {
+        val ll = l
+        r.blockMap(ll - _)
+      }
+      def *(r: M): M =
+        r.scalarMultiply(l)
+      def /(r: M): M = {
+        val ll = l
+        r.blockMap(l / _)
+      }
+    }
+  }
+}
+
+// must be top-level for Jackson to serialize correctly
+case class HailBlockMatrixMetadata(blockSize: Int, rows: Long, cols: Long)
+
+class HailBlockMatrix(val blocks: RDD[((Int, Int), BDM[Double])],
+  val blockSize: Int,
+  val rows: Long,
+  val cols: Long) extends Serializable {
+  type M = HailBlockMatrix
+
+  require(blocks.partitioner.isDefined)
+  require(blocks.partitioner.get.isInstanceOf[HailGridPartitioner])
+
+  val partitioner = blocks.partitioner.get.asInstanceOf[HailGridPartitioner]
+
+  def transpose(): M =
+    new HailBlockMatrix(new HailBlockMatrixTransposeRDD(this), blockSize, cols, rows)
+
+  def diagonal(): Array[Double] = {
+    require(this.rows == this.cols,
+      s"diagonal only works on square matrices, given ${ this.rows }x${ this.cols }")
+
+    def diagonal(block: BDM[Double]): Array[Double] = {
+      val length = math.min(block.rows, block.cols)
+      val diagonal = new Array[Double](length)
+      var i = 0
+      while (i < length) {
+        diagonal(i) = block(i, i)
+        i += 1
+      }
+      diagonal
+    }
+
+    this.blocks
+      .filter { case ((i, j), block) => i == j }
+      .map { case ((i, j), block) => (i, diagonal(block)) }
+      .collect()
+      .sortBy(_._1)
+      .map(_._2)
+      .fold(Array[Double]())(_ ++ _)
+  }
+
+  def add(other: M): M =
+    blockMap2(other, _ + _)
+
+  def subtract(other: M): M =
+    blockMap2(other, _ - _)
+
+  def pointwiseMultiply(other: M): M =
+    blockMap2(other, _ :* _)
+
+  def pointwiseDivide(other: M): M =
+    blockMap2(other, _ :/ _)
+
+  def multiply(other: M): M =
+    new HailBlockMatrix(new HailBlockMatrixMultiplyRDD(this, other), blockSize, rows, other.cols)
+
+  def multiply(r: BDM[Double]): M = {
+    require(this.cols == r.rows,
+      s"incompatible matrix dimensions: ${ this.rows }x${ this.cols } and ${ r.rows }x${ r.cols }")
+    multiply(HailBlockMatrix.from(this.blocks.sparkContext, r, this.blockSize))
+  }
+
+  def scalarAdd(i: Double): M =
+    this.blockMap(_ + i)
+
+  def scalarSubtract(i: Double): M =
+    this.blockMap(_ - i)
+
+  def scalarMultiply(i: Double): M =
+    this.blockMap(_ :* i)
+
+  def scalarDivide(i: Double): M =
+    this.blockMap(_ / i)
+
+  def vectorAddToEveryColumn(v: Array[Double]): M = {
+    require(v.length == this.rows, s"vector length, ${ v.length }, must equal number of matrix rows ${ this.rows }; v: $v, m: $this")
```

Comment:
I'm not sure I understood what you want. I added a comma after "rows". I don't want to delete the vector length since I think it's helpful to know how long the vector is if it doesn't match the number of rows.